contributing/LOCKING.md

```diff
@@ -0,0 +1,88 @@
+# Locking
+
+The `dstack` server supports SQLite and Postgres databases
+with two implementations of resource locking to handle concurrent access:
+
+* In-memory locking for SQLite.
+* DB-level locking for Postgres.
+
+## SQLite locking
+
+SQLite is missing efficient mechanisms to handle concurrent writes (e.g. select for update),
+so `dstack` implements in-memory resource-level locking.
+In-memory locking works correctly under the assumption that there is only one server instance (process),
+which is a `dstack` limitation when using SQLite.
+
+The in-memory locking is implemented via locksets.
+Locksets are Python sets that store IDs of locked resources.
+Concurrent access to locksets is guarded with asyncio locks:
+
+```python
+lock, locket = get_lockset("my_table")
+async with lock:
+    # select resource that is not in lockset
+    lockset.add(resource.id)
+try:
+    process_resource(resource)
+finally:
+    lockset.remove(resource.id)
+```
+
+Locksets are an optimization. One can think of them as per-resource-id locks that
+allow independent locking of different resources.
+
+## Postgres locking
+
+Postgres resource locking is implemented via standard SELECT FOR UPDATE.
+SQLAlchemy provides `.with_for_update()` that has not effect if SELECT FOR UPDATE is not supported as in SQLite.
+
+There are few places that rely on advisory locks as when generating unique resource names.
+
+## Working with locks
+
+Concurrency is hard. Below you'll find common patterns and gotchas when working with locks to make it a bit more manageable.
+
+**A task should acquire locks on resources it modifies**
+
+This is a common sense approach. An alternative could be the inverse: job processing cannot run in parallel with run processing, so job processing takes run lock. This indirection complicates things and is discouraged.
```

Comment:
Sorry, I didn't quite understand the example before actually looking at the run and job processing code :)
```suggestion
This is a common sense approach. An alternative could be the inverse: job processing cannot run in parallel with run processing, so job processing takes run lock. This indirection complicates things and is discouraged. In this example, run processing should take job lock instead.
```